openvdb_points/tools/AttributeGroup.cc

```diff
@@ -0,0 +1,112 @@
+///////////////////////////////////////////////////////////////////////////
+//
+// Copyright (c) 2015 Double Negative Visual Effects
+//
+// All rights reserved. This software is distributed under the
+// Mozilla Public License 2.0 ( http://www.mozilla.org/MPL/2.0/ )
+//
+// Redistributions of source code must retain the above copyright
+// and license notice and the following restrictions and disclaimer.
+//
+// *     Neither the name of Double Negative Visual Effects nor the names
+// of its contributors may be used to endorse or promote products derived
+// from this software without specific prior written permission.
+//
+// THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
+// "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
+// LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR
+// A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
+// OWNER OR CONTRIBUTORS BE LIABLE FOR ANY INDIRECT, INCIDENTAL,
+// SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
+// LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
+// DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
+// THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+// (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+// OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+// IN NO EVENT SHALL THE COPYRIGHT HOLDERS' AND CONTRIBUTORS' AGGREGATE
+// LIABILITY FOR ALL CLAIMS REGARDLESS OF THEIR BASIS EXCEED US$250.00.
+//
+///////////////////////////////////////////////////////////////////////////
+//
+/// @file AttributeGroup.cc
+///
+/// @authors Dan Bailey
+
+
+#include <openvdb_points/tools/AttributeGroup.h>
+
+
+namespace openvdb {
+OPENVDB_USE_VERSION_NAMESPACE
+namespace OPENVDB_VERSION_NAME {
+namespace tools {
+
+
+////////////////////////////////////////
+
+// GroupAttributeArray implementation
+
+
+GroupAttributeArray::GroupAttributeArray(size_t n, const ValueType& uniformValue)
+    : TypedAttributeArray<GroupType, NullAttributeCodec<GroupType> >(n, uniformValue)
+{
+    this->setGroup(true);
+}
+
+
+GroupAttributeArray::GroupAttributeArray(const TypedAttributeArray& array, const bool decompress)
+        : TypedAttributeArray<GroupType, NullAttributeCodec<GroupType> >(array, decompress)
+{
+    this->setGroup(true);
+}
+
+
+void GroupAttributeArray::setGroup(bool state)
+{
+    if (state) mFlags |= Int16(GROUP);
+    else mFlags &= ~Int16(GROUP);
+}
+
+
+////////////////////////////////////////
+
+// GroupHandle implementation
+
+
+bool GroupHandle::get(Index n) const
+{
+    return (mArray.get(n) & mBitMask) == mBitMask;
```

Comment:
I admit this operation looks a bit strange, but it needs to ensure that all bits in the mask are matched. 

Here's an example that you want to allow:

Array: 0 1 1 0 0 1 0 0
BitMask: 0 1 1 0 0 0 0 0

Array & BitMask: 0 1 1 0 0 0 0 0 (correctly evaluates to true)

and one you don't:

Array: 0 1 1 0 0 0 0 0
BitMask: 0 1 1 0 0 1 0 0

Array & BitMask: 0 1 1 0 0 0 0 0 (incorrectly evaluates to true)

Perhaps there's a better way of expressing this.